### ISSUE TITLE
Application Cookie, segregating by domain

### ISSUE THREAD

**Issue by dazinator** [aspnet/Identity#968]
TLDR: Cannot share identity cookies accross domains: Repro here: https://github.com/dazinator/AspNetCore-Identity-MultiDomain/blob/master/README.md

My website is hosted on the following domains:

127.0.0.1   app.foo.com
127.0.0.1   app.bar.com
127.0.0.1   another.bar.com

I would like to share the identity cookie accross a top level domain i.e ".bar.com". 
In other words, I'd like for a user to login on `app.bar.com` and navigate to `another.bar.com` and still be logged in as the cookie is still present. 
However if they navigate to `app.foo.com` they will have to log in seperately.

By default, asp Identity uses a different cookie for each different hostname. 

I have been able to achieve this prior to having Identity in the mix, i.e I could call the Cookie Authentication Middleware directly multiple times:

```
 cookieOptions.Domain = ".foo.com";
 app.UseCookieAuthentication(cookieOptions);

 cookieOptions.Domain = ".bar.com";
 app.UseCookieAuthentication(cookieOptions);
```

And that gave me the behaviour I wanted.

But now that I have added Identity into the mix, how do I configure the application cookie for Identity to do this?

I have tried this which doesn't work:

```

 // localhost;
            services.AddIdentity<ApplicationUser, IdentityRole>(a =>
            {
                var appCookie = a.Cookies.ApplicationCookie;
                appCookie.ExpireTimeSpan = TimeSpan.FromMinutes(120);
                appCookie.AutomaticChallenge = true;
                appCookie.AutomaticAuthenticate = true;
                appCookie.CookieName = "BarIdentity";
                appCookie.LoginPath = "/Account/Login";
                appCookie.AuthenticationScheme = new IdentityOptions().Cookies.ApplicationCookie.AuthenticationScheme;
            });

            // bar.com
            services.AddIdentity<ApplicationUser, IdentityRole>(a =>
            {
                var appCookie = a.Cookies.ApplicationCookie;
                appCookie.CookieDomain = ".bar.com";
                appCookie.ExpireTimeSpan = TimeSpan.FromMinutes(120);
                appCookie.AutomaticChallenge = true;
                appCookie.AutomaticAuthenticate = true;
                appCookie.CookieName = "BarIdentity";
                appCookie.LoginPath = "/Account/Login";
                appCookie.AuthenticationScheme = new IdentityOptions().Cookies.ApplicationCookie.AuthenticationScheme + ".Bar";
            })
          .AddEntityFrameworkStores<ApplicationDbContext>()
          .AddDefaultTokenProviders();

```

I have also tried adding Identity once, and then calling 

```

 cookieOptions.Domain = ".app.com";
 app.UseCookieAuthentication(cookieOptions);

 cookieOptions.Domain = ".bar.com";
 app.UseCookieAuthentication(cookieOptions);
```

Didn't work either.

**Comment by dazinator** [comment]
Repro here: https://github.com/dazinator/AspNetCore-Identity-MultiDomain/blob/master/README.md

**Comment by blowdart** [comment]
Why you say doesn't work, what do you actually mean?

**Comment by dazinator** [comment]
I mean it doesn't work, in that, due to the fact that Identity registers `IdentityOptions` as a singleton, and due to the fact that it uses this class to configure its cookies (like the application sign in cookie), if you need to configure multiple distinct application sign in cookies (for example one for *.foo.com and one for *.bar.com) - you can't do this. 

The illustration, is that if you have a site that is accessed on:

- one.foo.com
- two.foo.com
- one.bar.com
- two.bar.com

It's not currently possible to set up Identity so that it can issue a sign in cookie for domain '.bar.com` and another sign in cookie for domain '.foo.com` because when you `AddIdentity()` you can only ever configure a single `IdentityOptions` which means you can only ever configure a single Application cookie, which means setting a single domain for that single cookie.

**Comment by HaoK** [comment]
Yes, identity does not expect to be multi instance, unlike cookies which is.
